dgisim/tests/test_dices.py

```diff
@@ -167,101 +168,438 @@ def test_just_satisfy_typical_all_omni(self):
+                    )
+
+                    if element_number >= omni_number:
+                        expected = ActualDices(dices={element: omni_number})
+                    else:
+                        expected = None
+                    self.assertEqual(actual, expected)
+
+    def test_smart_selection_any(self):
+        """test one Element can fill ANY"""
+        element: Element = Element.GEO
+        for any_number in (2, 3, 5):
+            for element_number in (2, 3, 5):
+                with self.subTest(any_number=any_number, element_number=element_number):
+                    input_actual_dices: ActualDices = ActualDices(dices={element: element_number})
+                    input_abstract_dices: AbstractDices = AbstractDices(
+                        dices={Element.ANY: any_number}
+                    )
+                    input_precedence: list[set[Element]] = []
+
+                    actual = input_actual_dices.smart_selection(
+                        requirement=input_abstract_dices,
+                        game_state=None,
+                        local_precedence_arg=input_precedence
+                    )
+                    if element_number >= any_number:
+                        expected = ActualDices(dices={element: any_number})
+                    else:
+                        expected = None
+
+                    self.assertEqual(actual, expected)
+
+    def test_smart_selection_all_any(self):
+        """test one die of every Element can fill ANY in proper order"""
+        for number_of_any in range(1, len(Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY) + 1):
+            with self.subTest(number_of_any=number_of_any):
+                input_actual_dices: ActualDices = ActualDices(
+                    dices={el: 1 for el in Dices._ELEMENTS}
+                )
+                input_abstract_dices: AbstractDices = AbstractDices(
+                    dices={Element.ANY: number_of_any}
+                )
+                input_precedence: list[set[Element]] = [
+                    {el} for el in Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY
+                ]
+
+                actual: Optional[ActualDices] = input_actual_dices.smart_selection(
+                    requirement=input_abstract_dices, game_state=None, local_precedence_arg=input_precedence
+                )
+
+                if Dices._NUMBER_OF_ELEMENTS >= number_of_any:
+                    expected: Optional[ActualDices] = ActualDices(
+                        {el: 1 for el in Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY[-number_of_any:]}
+                    )
+                else:
+                    expected = None
+
+                self.assertEqual(actual, expected)
+
+    def test_smart_selection_all_any_plus_one_omni_actual(self):
+        """test one die of every element from a to b, b-a+1 ANY
+        a<b"""
+        elements_and_omni = [Element.OMNI] + list(Dices._ELEMENTS_BY_DECREASING_GLOBAL_PRIORITY)
```

Comment:
not very important here, but again, you can concat tuples like this `tuple_a + tuple_b`

Reply:
done